src/main/java/com/zaxxer/hikari/pool/HikariPool.java

```diff
@@ -631,13 +631,24 @@ public void run()
          }
 
          logPoolState("Before cleanup ");
-         for (PoolBagEntry bagEntry : connectionBag.values(STATE_NOT_IN_USE)) {
+         boolean evicted = false;
+         List<PoolBagEntry> bag = connectionBag.values(STATE_NOT_IN_USE);
+         for (PoolBagEntry bagEntry : bag) {
+            if (bagEntry.evicted) {
+               closeConnection(bagEntry, "(connection evicted)");
+               evicted = true;
+            }
+         }
+         if (evicted) {
+        	//get fresh bag
+            bag = connectionBag.values(STATE_NOT_IN_USE);
+         }
+         int removable = bag.size() - config.getMinimumIdle();
+         for (PoolBagEntry bagEntry : bag) {
             if (connectionBag.reserve(bagEntry)) {
-               if (bagEntry.evicted) {
-                  closeConnection(bagEntry, "(connection evicted)");
-               }
-               else if (idleTimeout > 0L && clockSource.elapsedMillis(bagEntry.lastAccess, now) > idleTimeout) {
+               if (removable > 0 && idleTimeout > 0L && clockSource.elapsedMillis(bagEntry.lastAccess, now) > idleTimeout) {
                   closeConnection(bagEntry, "(connection passed idleTimeout)");
+                  removable--;
```

Comment:
call to poolElf.quietlyCloseConnection() is async in closeConnection()
I considered not to decrement removables in case of evicted ..but that is problematic as removable is not calculated correctly with some evicted connection in bag